### ISSUE TITLE
(strings-lazy-pp) Assertion error at src/./expr/node.h:1107

### ISSUE THREAD

**Issue by rainoftime** [CVC4/CVC4#5938]
cvc4 d278cfe019534f876
~~~~
(set-option :strings-lazy-pp false)
(declare-fun i1 () Int)
(declare-fun seq7 () (Seq Int))
(declare-fun seq9 () (Seq Int))
(declare-fun str4 () String)
(declare-fun str7 () String)
(declare-fun str8 () String)
(declare-fun str9 () String)
(assert (= str9 (str.substr str9 0 (str.to_int str9))))
(push)
(assert (= (str.substr str4 0 i1) (str.substr str9 0 (str.to_int str9))))
(assert (= str9 (str.substr str9 i1 1)))
(assert (= str9 (str.substr str9 (- 1 i1) 1)))
(assert (or (seq.suffixof seq9 seq7) (and (= str9 (str.substr str9 1 1)) (= (str.substr str8 1 i1) (str.substr str7 1 1)))))
(check-sat)
(assert (or (seq.suffixof (seq.unit 0) seq9) (= (str.substr str4 i1 1) (str.substr str9 0 (str.to_int str9)))))
(push)
(assert (or (seq.contains seq9 (seq.unit 0)) (= (str.rev str9) (str.substr (str.substr str9 0 (str.to_int str8)) (- 0 i1 1) (- i1)))))
(check-sat)
~~~~

~~~~
cvc4 -q -i delta.out.smt2 
sat
Fatal failure within CVC4::NodeTemplate<ref_count>::NodeTemplate(const CVC4::NodeTemplate<ref_count>&) [with bool ref_count = false] at /home/CVC4/src/./expr/node.h:1107
Check failure

 d_nv->d_rc > 0
TNode constructed from TNode with rc == 0
Aborted
~~~~

**Comment by ajreynol** [comment]
Might be related to justification heuristic, backtrace:
```
Fatal failure within CVC4::NodeTemplate<ref_count>::NodeTemplate(const CVC4::NodeTemplate<ref_count>&) [with bool ref_count = false] at /space/ajreynol/cvc4-ajr/src/./expr/node.h:1107
Check failure

 d_nv->d_rc > 0
TNode constructed from TNode with rc == 0

Program received signal SIGABRT, Aborted.
0x00000000018884d5 in raise ()
(gdb) bt
#0  0x00000000018884d5 in raise ()
#1  0x00000000004032f9 in abort ()
#2  0x000000000090d9d9 in CVC4::FatalStream::~FatalStream (this=0x7fffffffb90e, __in_chrg=<optimized out>) at /space/ajreynol/cvc4-ajr/src/base/check.cpp:33
#3  0x0000000000462102 in CVC4::NodeTemplate<false>::NodeTemplate (this=0x7fffffffbb08, e=...) at /space/ajreynol/cvc4-ajr/src/./expr/node.h:1107
#4  0x00000000012366b8 in CVC4::decision::JustificationHeuristic::getNextThresh (this=0x268f6a0, stopSearch=@0x7fffffffbd87: false, threshold=0)
    at /space/ajreynol/cvc4-ajr/src/decision/justification_heuristic.cpp:108
#5  0x000000000123613f in CVC4::decision::JustificationHeuristic::getNext (this=0x268f6a0, stopSearch=@0x7fffffffbd87: false)
    at /space/ajreynol/cvc4-ajr/src/decision/justification_heuristic.cpp:77
#6  0x000000000123424b in CVC4::DecisionEngine::getNext (this=0x268f630, stopSearch=@0x7fffffffbd87: false) at /space/ajreynol/cvc4-ajr/src/decision/decision_engine.cpp:80
#7  0x0000000000c14916 in CVC4::prop::TheoryProxy::getNextDecisionEngineRequest (this=0x26901e0, stopSearch=@0x7fffffffbd87: false)
    at /space/ajreynol/cvc4-ajr/src/prop/theory_proxy.cpp:134
#8  0x0000000000bc4f3e in CVC4::Minisat::Solver::pickBranchLit (this=0x26917c0) at /space/ajreynol/cvc4-ajr/src/prop/minisat/core/Solver.cc:870
#9  0x0000000000bc9a41 in CVC4::Minisat::Solver::search (this=0x26917c0, nof_conflicts=25) at /space/ajreynol/cvc4-ajr/src/prop/minisat/core/Solver.cc:1796
#10 0x0000000000bca04e in CVC4::Minisat::Solver::solve_ (this=0x26917c0) at /space/ajreynol/cvc4-ajr/src/prop/minisat/core/Solver.cc:1891
#11 0x0000000000bee317 in CVC4::Minisat::SimpSolver::solve_ (this=0x26917c0, do_simp=false, turn_off_simp=false) at /space/ajreynol/cvc4-ajr/src/prop/minisat/simp/SimpSolver.cc:155
#12 0x0000000000bec421 in CVC4::Minisat::SimpSolver::solve (this=0x26917c0, do_simp=true, turn_off_simp=false) at /space/ajreynol/cvc4-ajr/src/./prop/minisat/simp/SimpSolver.h:228
#13 0x0000000000beaf82 in CVC4::prop::MinisatSatSolver::solve (this=0x25d3410) at /space/ajreynol/cvc4-ajr/src/prop/minisat/minisat.cpp:192
#14 0x0000000000bf7b05 in CVC4::prop::PropEngine::checkSat (this=0x268f5a0) at /space/ajreynol/cvc4-ajr/src/prop/prop_engine.cpp:346
#15 0x00000000004fa20d in CVC4::smt::SmtSolver::checkSatisfiability (this=0x25f15d0, as=..., assumptions=..., inUnsatCore=true, isEntailmentCheck=false)
    at /space/ajreynol/cvc4-ajr/src/smt/smt_solver.cpp:163
#16 0x00000000004d4e85 in CVC4::SmtEngine::checkSatInternal (this=0x25d54f0, assumptions=..., inUnsatCore=true, isEntailmentCheck=false)
    at /space/ajreynol/cvc4-ajr/src/smt/smt_engine.cpp:969
#17 0x00000000004d483e in CVC4::SmtEngine::checkSat (this=0x25d54f0, assumption=..., inUnsatCore=true) at /space/ajreynol/cvc4-ajr/src/smt/smt_engine.cpp:913
#18 0x00000000004d4722 in CVC4::SmtEngine::checkSat (this=0x25d54f0) at /space/ajreynol/cvc4-ajr/src/smt/smt_engine.cpp:898
#19 0x0000000000444841 in CVC4::api::Solver::checkSat (this=0x25aae70) at /space/ajreynol/cvc4-ajr/src/api/cvc4cpp.cpp:4829
#20 0x00000000004b4d0b in CVC4::CheckSatCommand::invoke (this=0x2a66ce0, solver=0x25aae70, sm=0x25f1e80) at /space/ajreynol/cvc4-ajr/src/smt/command.cpp:421
#21 0x00000000004b39fe in CVC4::Command::invoke (this=0x2a66ce0, solver=0x25aae70, sm=0x25f1e80, out=...) at /space/ajreynol/cvc4-ajr/src/smt/command.cpp:208
#22 0x000000000040cd78 in CVC4::main::solverInvoke (solver=0x25aae70, sm=0x25f1e80, cmd=0x2a66ce0, out=0x259dc20 <std::cout>)
    at /space/ajreynol/cvc4-ajr/src/main/command_executor.cpp:218
#23 0x000000000040c50c in CVC4::main::CommandExecutor::doCommandSingleton (this=0x25d1200, cmd=0x2a66ce0) at /space/ajreynol/cvc4-ajr/src/main/command_executor.cpp:129
#24 0x000000000040c3d2 in CVC4::main::CommandExecutor::doCommand (this=0x25d1200, cmd=0x2a66ce0) at /space/ajreynol/cvc4-ajr/src/main/command_executor.cpp:105
#25 0x00000000004078ba in CVC4::main::CommandExecutor::doCommand (this=0x25d1200, cmd=...) at /space/ajreynol/cvc4-ajr/src/./main/command_executor.h:73
#26 0x00000000004065ce in runCvc4 (argc=4, argv=0x7fffffffdf28, opts=...) at /space/ajreynol/cvc4-ajr/src/main/driver_unified.cpp:447
#27 0x000000000040bc7a in main (argc=4, argv=0x7fffffffdf28) at /space/ajreynol/cvc4-ajr/src/main/main.cpp:52
```

**Comment by nafur** [comment]
While the assertion failure is resolved by #5993, a minimized version of the input still triggers an error:
```
$ cat out.smt2 
(set-option :strings-lazy-pp false)
(declare-fun s () (Seq Int))
(declare-fun t () String)
(push)
(assert (or (seq.contains s (seq.unit 0)) (= (str.rev t) (str.substr t 0 (str.to_int t)))))
(check-sat)
$ bin/cvc4 -i --strings-exp out.smt2 
(error "Expecting a string term as the first argument in 'STRING_TO_CODE'")
cvc4: /src/prop/minisat/core/Solver.cc:2094: void CVC4::Minisat::Solver::pop(): Assertion `decisionLevel() == 0' failed.
Aborted (core dumped)
```
Note that the assertion from minisat vanishes, if the `push` is removed from the input.
